### PR TITLE
Some basic events for changes

### PR DESCRIPTION
**Pull request by chucker**
Includes a basic sample (that only logs to console for now).

Stuff like listening for a deleted key isn't currently implemented.

### CODE REVIEW

File: src/Blazored.LocalStorage/LocalStorageService.cs [hunk 1]
```diff
@@ -20,7 +20,16 @@ public Task SetItem(string key, object data)
...
 
-            return _jSRuntime.InvokeAsync<object>("Blazored.LocalStorage.SetItem", key, Json.Serialize(data));
+            RaiseOnChanging(key, data, out ChangingEventArgs e);
+
+            if (e.Cancel)
+                return Task.CompletedTask;
+
+            var jsResult = _jSRuntime.InvokeAsync<object>("Blazored.LocalStorage.SetItem", key, Json.Serialize(data));
+
+            RaiseOnChanged(key, e.OldValue, data);
```
Comment (kswoll) [inline src/Blazored.LocalStorage/LocalStorageService.cs]: I'm a little worried that this might not work as expected?  Since `InvokeAsync` above is not being awaited, `RaiseOnChanged` will actually (potentially) fire before `SetItem` has completed.

Comment (chucker) [reply src/Blazored.LocalStorage/LocalStorageService.cs]: I think you're right

Comment (chucker) [reply src/Blazored.LocalStorage/LocalStorageService.cs]: Actually, I'm confused about the contract of `public Task SetItem(string key, object data)`.

* shouldn't it be `async`?
* shouldn't it return `Task<object>` instead of `Task`? Currently, it seems to return the `jsRuntime` return value as a `Task`?

Comment (kswoll) [reply src/Blazored.LocalStorage/LocalStorageService.cs]: A method only needs to be `async` if it uses `await` -- this makes the compiler create the whole async state machine (and thus should be avoided if not using `await`).  Since the method isn't awaiting, it can just return the `Task` outright (and `Task<T>` is a subclass of `Task` which is why you can return the former as the latter)

It _could_ return `Task<object>` but the only reason we're passing a type to `Invoke` at all, is because there isn't an overload that doesn't take a generic type argument -- the JS function doesn't return a value anyway.

tl;dr you should be able to make the method `async` and change the code to:

```
await _jSRuntime.InvokeAsync<object>("Blazored.LocalStorage.SetItem", key, Json.Serialize(data));

RaiseOnChanged(key, e.OldValue, data);
```

And then not return anything from the method at all.

Comment (kswoll) [reply src/Blazored.LocalStorage/LocalStorageService.cs]: Ah, saw your update.  Yeah, you can just discard the return value (`_`) entirely, and note that technically your final `return;` statement is now unnecessary.

Comment (chucker) [reply src/Blazored.LocalStorage/LocalStorageService.cs]: I basically did that. Check my later commits.

File: src/Blazored.LocalStorage/LocalStorageService.cs [hunk 2]
```diff
@@ -15,12 +15,21 @@ public LocalStorageService(IJSRuntime jSRuntime)
...
+            RaiseOnChanging(key, data, out ChangingEventArgs e);
+
+            if (e.Cancel)
+                return;
+
+            _ = await _jSRuntime.InvokeAsync<object>("Blazored.LocalStorage.SetItem", key, Json.Serialize(data));
+
+            RaiseOnChanged(key, e.OldValue, data);
+
+            return;
```
Comment (chrissainty) [inline src/Blazored.LocalStorage/LocalStorageService.cs]: Echoing what @kswoll said, this `return;` can go.

File: src/Blazored.LocalStorage/LocalStorageService.cs [hunk 3]
```diff
@@ -15,12 +15,21 @@ public LocalStorageService(IJSRuntime jSRuntime)
...
             if (string.IsNullOrEmpty(key))
                 throw new ArgumentNullException(nameof(key));
 
-            return _jSRuntime.InvokeAsync<object>("Blazored.LocalStorage.SetItem", key, Json.Serialize(data));
+            RaiseOnChanging(key, data, out ChangingEventArgs e);
+
+            if (e.Cancel)
+                return;
+
+            _ = await _jSRuntime.InvokeAsync<object>("Blazored.LocalStorage.SetItem", key, Json.Serialize(data));
```
Comment (chrissainty) [inline src/Blazored.LocalStorage/LocalStorageService.cs]: There is no need to use a discard here. You can just await the call.

Comment (chucker) [reply src/Blazored.LocalStorage/LocalStorageService.cs]: (The IDE suggested it, because you were previously returning the value.)
